### ISSUE TITLE
Values only as local variables, instead of class variables in the contents. (or also: dealing with collections)

### ISSUE THREAD

**Issue by 0xBEEEF** [kaitai-io/kaitai_struct#379]
The following problem/requirement: Task is to create an object wrapper using a binary. Let's take as an example one of the supplied examples, for example the file [java_class.ksy](https://github.com/kaitai-io/kaitai_struct_formats/blob/9a38c7d4c5058f7302e0a3272adfde78f6e62a07/executable/java_class.ksy). This file is relatively complex in structure and consists of some repeated elements. In order to know that this is a repetition, a "number" is required in the file. This "number" is for me in the object model in a certain way but relatively irrelevant. It even confuses in part, because the block that is repeatedly mapped as an object construct, since it is a collection of a subtype or an array. The number of objects is not in this object, but in the object above it. This means that I don't really need this "number" in the object above. Even if the serialization is available at a later time, then I would have to keep this "number" synchronized with the "collection". In the object-oriented world, however, this is of course a no-go because the collection itself knows how many elements it has or how long it will be. This is of course completely different per programming language, but in principle it is in a way comparable.

If we look at the file java_class.ksy again, for example, we see that we have several "number" elements here. In other words, you could reduce the result if you did not have this problem. 

```
  - id: constant_pool_count
    type: u2
  - id: constant_pool
    type: constant_pool_entry
    repeat: expr
    repeat-expr: constant_pool_count - 1
```

I would therefore like to ask the general question whether there will be changes in the future, or whether certain elements will be read but not made available to the class as an independent property.

What I would still see as a possible solution would be to provide a kind of separate subspecies "Collection" for repeated elements. This is based on a count and contains a child type, which is repeated X-times. The collection would have a managed property "number", which would depend on the contained elements. This would also solve this problem in the subsequent serialization process. Sooner or later, there'd have to be a lion in here anyway.

**Comment by KOLANICH** [comment]
#88

**Comment by GreyCat** [comment]
@0xBEEEF Honestly, I failed to understand almost every single sentence. Could you rephrase your question, so it would be clearer?

> Task is to create an object wrapper using a binary.

What is "object wrapper", what is "binary", what's this all about?

> This file is relatively complex in structure and consists of some repeated elements.

Sure.

> In order to know that this is a repetition, a "number" is required in the file.

Um, what? If you need repetition of a particular attribute, you just do repetition in ksy — i.e. as you pointed out, using `repeat: XXX` and `repeat-XXX: YYY` clauses. It's still unclear what that "number" that you mentioned so many times is.

> This "number" is for me in the object model in a certain way but relatively irrelevant.

Ok, so this "number" is for you, and what it is for everyone else? What does "in the object model in a certain way" means? Why are you mentioning it anyway, if it is "irrelevant"?

Etc.

**Comment by 0xBEEEF** [comment]
By Object Wrapper I mean an object model based on a binary file.

By irrelevant I mean that in my opinion it makes no sense to store any count-values for repeated ranges in the generated classes if they are only used for creating collections, because a collection itself knows how many values it contains. From my point of view, it makes no sense to cancel the countdate values of the original file in the generated classes. Instead, they should be wrapped in their own collection class, and therefore depend dynamically on the number of objects in the collection. What I also meant was that if there is support for serialization at some point, it doesn't make sense to manually set the count-values that are included in the generated classes. This means that if you serialize the created object model again, you also have to update all counts. Instead, you could easily work with a collection class that could do all this.

**Comment by GreyCat** [comment]
Ok, so, basically, this whole thing sums up to the fact that for simple stuff like:

```yaml
seq:
  - id: num_entries
    type: u4
  - id: entries
    type: entry
    repeat: expr
    repeat-expr: num_entries
```

... you don't want to have store to `num_entries` in a generated class, only `entries` itself?

This kind of contradicts with the core idea of KS reflecting the data stored in a binary format, not doing any transformations on the fly. And, unfortunately, things get much more complicated in more complex cases, i.e. when `repeat-expr` values are fetched from another types, or are a result of complex calculation.

**Comment by 0xBEEEF** [comment]
Yes, that's exactly what the requirement was originally meant to be! I'm sorry for making a little misunderstanding. Unfortunately, I assumed a few things and didn't write them down. That's what happens when you deal intensively with a subject.

It is good to know that "Katai.struct" only provides the basic skeleton of the binary file. So far, I have assumed that it is a complete object mapper, as you know it from other programming languages. That was a misunderstanding on my part.

Sorry again for the bad description. That shouldn't have happened.

**Comment by cugu** [comment]
I'll close this issue as is seems to be resolved. Please reopen if you feel this is wrong.
